src/main.js

```diff
@@ -1,144 +1,175 @@
+        const { hasValidScope } = require('./validate/editor')
+
         const { fixOnSave, scopes } = atomConfig
         if (hasValidScope(editor, scopes) && fixOnSave) {
           await this.fixJob(true)
         }
       })
     }))
 
+    // Subscribe to Debug command
     this.subscriptions.add(atom.commands.add('atom-text-editor', {
       'linter-eslint:debug': async () => {
-        loadDeps()
-        const debugString = await debug.report()
-        const notificationOptions = { detail: debugString, dismissable: true }
+        const { report } = require('./debug')
+
+        const debugReport = await report()
+        const notificationOptions = { detail: debugReport, dismissable: true }
         atom.notifications.addInfo('linter-eslint debugging information', notificationOptions)
       }
     }))
 
+    // Subscribe to Fix File command
     this.subscriptions.add(atom.commands.add('atom-text-editor', {
       'linter-eslint:fix-file': async () => {
         await this.fixJob()
       }
     }))
 
+    // Add context menu entry for Fix File
     this.subscriptions.add(atom.contextMenu.add({
       'atom-text-editor:not(.mini), .overlayer': [{
         label: 'ESLint Fix',
         command: 'linter-eslint:fix-file',
         shouldDisplay: (evt) => {
           const activeEditor = atom.workspace.getActiveTextEditor()
-          if (!activeEditor) {
-            return false
-          }
-          // Black magic!
+          if (!activeEditor) return false
+
+          // Some scary voodoo black magic! Atom v1.19.0+
           // Compares the private component property of the active TextEditor
-          //   against the components of the elements
+          //   against the components property of the TextEditor DOM elements
           const evtIsActiveEditor = evt.path.some(elem =>
-            // Atom v1.19.0+
             (elem.component && activeEditor.component
               && elem.component === activeEditor.component))
+
           // Only show if it was the active editor and it is a valid scope
           const { scopes } = atomConfig
+          const { hasValidScope } = require('./validate/editor')
+
           return evtIsActiveEditor && hasValidScope(activeEditor, scopes)
         }
       }]
     }))
 
-    scheduleIdleTasks()
+    // Load miscellaneous idle tasks and potential settings migrations
+    if (!atom.inSpecMode()) {
+      this.subscriptions.add(makeIdleRunner()(getMigrations()))
+      this.subscriptions.add(makeIdleRunner()(idleTasks))
+    }
   },
 
   deactivate() {
-    idleCallbacks.forEach(callbackID => window.cancelIdleCallback(callbackID))
-    idleCallbacks.clear()
-    if (worker) {
-      // If the helpers module hasn't been loaded then there was no chance a
-      // worker was started anyway.
-      worker.task.kill(true)
-    }
+    const { task } = require('./worker-manager')
+    if (task) task.kill(true)
     this.subscriptions.dispose()
   },
 
   provideLinter() {
+    const {
```

Comment:
If you are going to abuse `require`, _own_ it and just `require` this stuff before it is used.

Reply:
That's about readability. Truly-inline requires can get pretty noisy. Once we reach the top of that function we know for certain we need it now. The one you chose to comment on is particularly notable for the fact that we also need to dig in for more nested data. 

```
showRule: require('./atom-config').atomConfig.showRule
```

The main point of the "inline" requiring is so that we can get all the imports inside of functions, thus avoiding Atom trying to parse the import files right away, and also making it possible for Atom to make snapshots for even faster loads. Once we are inside a function, it mostly doesn't matter if we import at the top of the function or later on (except for in the very first `activate` function, where we are just hopeful we manage to _somehow_ get at least 1 or 2 of them loaded before we get there. 

The point of using `require` instead of variables is 1.because its a lot easier to see what's actually going on when you don't have this out-of-nowhere function that magically makes unspecified names suddenly available that weren't usable before calling it.

And 2. It's easier to idly load, when you don't have to worry about whether the idle loading was in-time or not. We just say "load this when you can", and if it doesn't happen, so what. "We'll just grab this one ourselves, and you load the next one."

This method tries to treat it as close as possible to the more common `import`s. We're just importing at the top of the function instead of at the top of the file. 

The `require` function has a certain behavior, which includes caching. I don't think taking advantage of that behavior when we're pretty much _forced_ to use it in place of `import` qualifies as "abusing" it.

Reply:
> making it possible for Atom to make snapshots for even faster loads.

This is an _incredibly_ long term goal btw, snapshots are currently only done for packages bundled into Atom itself.

> except for in the very first activate function, where we are just hopeful we manage to somehow get at least 1 or 2 of them loaded before we get there.

😮 Uh.... that doesn't sound good _at all_, previously dependencies were only needed in commands that could only be triggered by the user, and even then they were ensured to be loaded by the time they were used. We should _never_ be "hopeful we manage to somehow get at least 1 or 2 of them loaded before we get there".

Reply:
Only hopeful in that we hope the scheduler beats us to it while we are in the middle of something else anyway. Either way, when we actually get to where the code is needed, we will get that code. There is nothing questionable about that.